### PR TITLE
Handle nested gitignores

### LINKED ISSUE #487: .gitignore and unix path separator doesn't work on Windows
I'm trying to use this on a Hugo website project that uses Pagefind for search, which has to run after Hugo, which means I can't use Hugo's built-in test server. When I run `watchexec --print-events -r "yarn run test"`, watchexec seemingly ignores my .gitignore file and continuously rebuilds on changes to Hugo's `public` directory.

```.gitignore
# My .gitignore file.
node_modules/
resources/
public/
.hugo_build.lock

# YARN Modern
.pnp.*
.yarn/*
!.yarn/patches
!.yarn/plugins
!.yarn/releases
!.yarn/sdks
!.yarn/versions
```

When I add the flag `-i {public,public\\**,resources,resources\\**}` works for ignoring the public, resources and their contents (apparently Hugo or Pagefind triggers a modify event on the public folder itself), however using `/` instead of `\\` for cross-platform compatibility fails to work.

- Watchexec's version: `1.21.0`
- The OS you're using: `Windows 11`
- A sample command that you've run that has the issue: `watchexec --print-events -r 'yarn run test'`

### LINKED ISSUE #578: Nested .gitignore breaks all ignores
When using `watchexec` on a workspace that has folders with `.gitignore` files that contain `*` (which basically tells git to ignore this whole folder), it completely breaks ignores for the rest of the workspace. 

You can clone this repro repo to see it in action:
https://github.com/Cammisuli/watch-exec-repro

Steps to reproduce:
1) clone repo
1) Open in vscode with rust analyzer (this just enables that files in the `target/` directory is automatically changed)
1) run `watchexec --project-origin=.  -- echo '$WATCHEXEC_WRITTEN_PATH'`
1) change `src/main.rs` (add a comment or whatever)
1) See output
	* the expected output should be just `main.rs`, but we get everything in `target/` as well.

1) Stop the `watchexec` command
1) Delete `this-should-ignore-children/.gitignore` (or change the glob to something else)
1) run `watchexec --project-origin=.  -- echo '$WATCHEXEC_WRITTEN_PATH'`
1) change `src/main.rs`
1) Output is only `src/main.rs` 

- Watchexec's version
 ```watchexec --version
watchexec 1.22.2 (2023-03-22) 
build-date: 2023-03-22
release: 1.22.2
features: 
```
- The OS you're using
```
macOS 13.3.1 (22E261)
```
- A log with `-vvvv --log-file` 
[watchexec.2023-05-02T18-19-42Z.log](https://github.com/watchexec/watchexec/files/11376435/watchexec.2023-05-02T18-19-42Z.log)

I'll try and investigate this myself and post any findings.

### PR DESCRIPTION
**Pull request by Cammisuli**
## What it does
This changes the way that filter handles `GitigoreBuilder` and `Gitignore` by allowing the creation of multiple instances depending on the origin of a gitignore file.

This uses a trie internally so that deeply nested files can be associated with parent `Gitignore` instances. 

Nested ignores are created based off their parent so that ignore inheritance is working as well.

fixes #578
fixes #487

TODO 

- [x] Write tests for nested ignores
- [x] Test with a compiled binary on macOS and Windows

### CODE REVIEW

**Review by passcod — APPROVED** [review]
That looks really good! And you even managed to not break API, I think.

**Comment by Cammisuli** [comment]
I tried to make sure that the API didnt break. All the function arguments and returns are still the same as before 🙂

**Comment by Cammisuli** [comment]
I found some issues on Windows. So I'm going to fix that before marking this as ready 🙂

**Comment by Cammisuli** [comment]
So it turns out that ignores never really worked in Windows, and the last bit of changes fixes that. (I've tested this on a Windows 11 machine myself to verify)

The biggest change for Windows is to use the [`dunce`](https://crates.io/crates/dunce) crate to handle Windows UNC paths (coming from `canonicalize` functions calls). This basically converts `\\?\C:\temp` paths to `C:\temp` paths so that things are a little bit more sensible. 

I had to remove the Windows path to unix path calls in the glob file so that Windows paths are properly used. 

I'm not too familiar with the public cli arguments, but if you'd like me to test some additional use cases, I'm up for that.

**Comment by passcod** [comment]
Hmmm, I previously used dunce but replaced it with the tokio canonicalize because dunce's is not async, and it would cause some bottlenecks. That said, I see the `simplified` function doesn't do any IO, so if that works for our purposes, fine with it.

**Review by passcod — APPROVED** [review]

**Comment by passcod** [comment]
No worries about the CLI, but if you could check out the clippy failures before I merge?

**Comment by Cammisuli** [comment]
I took the liberty to fix the `embed_resource` errors that came with the update to 2.x from dependabot and followed the migration guide here:
https://github.com/nabijaczleweli/rust-embed-resource#migration

**Comment by passcod** [comment]
Thanks a lot for this! I'll get it released sometime this weekend

**Comment by passcod** [comment]
Released in v1.22.3

File: crates/ignore-files/src/filter.rs [hunk 1]
```diff
@@ -75,108 +90,173 @@ impl IgnoreFilter {
...
+
 		for (file, content) in files_contents.into_iter().flatten() {
 			let _span = trace_span!("loading ignore file", ?file).entered();
+
+			let applies_in = get_applies_in_path(origin, Some(&file));
+
+			let parent_ignore = ignores_trie
+				.get_ancestor_value(&applies_in.display().to_string())
+				// unwrap will always succeed because we created an entry with the root of the origin
+				.unwrap();
```
Comment (kflansburg) [inline crates/ignore-files/src/filter.rs:116]: I'm encountering this `unwrap` after updating to `ignore-files` `1.3.0` in previously-working code. This is what I was using before:

```rust
    let (ignores, errors) = ignore_files::from_origin(".").await;
    let mut filt = ignore_files::IgnoreFilter::new([origin](std::env::current_dir()?), &ignores).await?;
```

And it looks like the issue is fixed by using an absolute path in `from_origin`:

```rust
    let origin = std::env::current_dir()?;
    let (ignores, errors) = ignore_files::from_origin(&origin).await;
    let mut filt = ignore_files::IgnoreFilter::new(origin, &ignores).await?;
```

It seems that there was a breaking change to the API where you could previously pass `[IgnoreFile]` with relative paths to `IgnoreFilter::new`, but now this does not work with the new radix trie implementation. 

Anyway, I've fixed my issue, but just wanted to let you know and document for others who may run into the same thing.

Comment (passcod) [reply crates/ignore-files/src/filter.rs:116]: Ah, thanks for the report! I'll get that fixed up, release a .1, and yank the current version (in this order).

Comment (Cammisuli) [reply crates/ignore-files/src/filter.rs:116]: oh shoot, sorry about that. @passcod let me know if you want me to fix it, if you're too busy.

Comment (passcod) [reply crates/ignore-files/src/filter.rs:116]: I'd appreciate it @Cammisuli ! I'm really flat out elsewhere currently 😓

Comment (passcod) [reply crates/ignore-files/src/filter.rs:116]: 1.3.0 yanked as promised now that 1.3.1 has been released.

**Pull request merged**